pytometry/tools/clustering/_flowsom.py

```diff
@@ -0,0 +1,226 @@
+    som: MiniSom
+    meta_class: np.ndarray
+    consensus_clustering: ConsensusClustering
+
+
+def flowsom_clustering(
+    adata: AnnData,
+    key_added: str = "clusters",
+    som_dim: tuple[int, int] = (10, 10),
+    sigma: float = 1.0,
+    learning_rate: float = 0.5,
+    batch_size: int = 500,
+    seed: int = 42,
+    weight_init: str = "random",
+    neighbourhood_function: str = "gaussian",
+    min_clusters: int = 2,
+    max_clusters: int = 10,
+    n_resamples: int = 100,
+    resample_frac: float = 0.9,
+    copy: bool = False,
+    verbose: bool = False,
+    agglomerative_clustering_kwargs: dict | None = None,
+    return_clustering_objs: bool = False,
+) -> AnnData | tuple[AnnData, ClusteringObjects]:
+    """Cluster cytometry data using FlowSOM.
+
+    Based on the original FlowSOM R package by Van Gassen et al. (2015)
+    (https://onlinelibrary.wiley.com/doi/full/10.1002/cyto.a.22625.) and the python
+    implementation found at https://github.com/Hatchin/FlowSOM.
+
+    Args:
+        adata (AnnData): annotated data matrix of shape (n_samples, n_features)
+        key_added (str, default="clusters"): key under which to add the cluster labels
+        som_dim (tuple[int, int], default=(10, 10)): dimensions of the SOM
+        sigma (float, default=1.0): radius of the different neighbourhoods in the SOM
+        learning_rate (float, default=0.5): learning rate of the SOM
+        batch_size (int, default=500): batch size for training the SOM
+        seed (int, default=42): random seed for reproducibility
+        weight_init (str, default="random"): weight initialization method, either
+        "random" or "pca". NOTE: if "pca" is chosen, the input data must be scaled to
+        zero mean and unit variance.
+        neighbourhood_function (str, default="gaussian"): neighbourhood function for
+        the SOM
+        min_clusters (int, default=2): minimum number of clusters to consider
+        max_clusters (int, default=10): maximum number of clusters to consider
+        n_resamples (int, default=100): number of resamples for consensus clustering
+        resample_frac (float, default=0.5): fraction of samples to resample for
+        copy (bool, default=False): whether to copy the AnnData object or modify it
+        verbose (bool, default=False)
+        agglomerative_clustering_kwargs (dict, default=None): keyword arguments for
+        sklearn.cluster.AgglomerativeClustering. If None, defaults to
+        {"metric": "euclidean", "linkage": "average"}.
+        return_clustering_objs: whether to return the clustering objects in addition to
+        the annotated data matrix
+
+    Returns:
+        AnnData | tuple[AnnData, ClusteringObjects]: annotated data matrix with cluster
+        labels added under `key_added`. If `return_clustering_objs` is True, a tuple
+        containing the annotated data matrix and the clustering objects is returned.
+
+    """
+    adata = adata.copy() if copy else adata
```

Comment:
change to `adata = adata if inplace else adata.copy()`

Reply:
Done